### PR TITLE
Add slurm support

### PR DESCRIPTION
**Pull request by ffinfo**
### Description

_Please explain the changes you made here._
Adding Slurm support to Queue
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
Trying to run our pipelines on a slurm cluster from an other institute
_Mention any issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below. 

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] All tests passing on Bamboo (including license tests!)

#### Form
- [ ] Rebased, squashed and reworded to produce a single commit (exemptions may be made -- sparingly) that is _uncluttered by excess lines_ (lines that just say "addressed review", "fixed" etc left over from squashing)
- [ ] Both the PR and the final rebased commit have a **concise** yet **descriptive** title (they are used when we compile version notes)
- [ ] Edited the README / documentation accordingly and got sign-off from support team (private code is exempt)

#### Review
- [ ] Suggest a reviewer or ask your team lead to suggest one
- [ ] Final (thumbsup) from the reviewer(s)

#### GATK4 Implications
- [ ] If your fix/change is applicable to GATK4 as well, and is reasonably small and self-contained (< ~50 lines or so), port the change to GATK4 and open a PR against https://github.com/broadinstitute/gatk or https://github.com/broadinstitute/gatk-protected as appropriate, or at least make a "best effort" attempt to do so.
- [ ] If your fix/change cannot yet be ported to GATK4 because the tool in question hasn't been ported yet, or has only been partially ported, or it would be difficult/burdensome to port the change, or you tried to port the change and failed, then add the ticket to our list of [GATK3 PRs to be eventually ported to GATK4](https://docs.google.com/document/d/1DjEHw57k5h0i8MZRGYPlQA3InvURKwQ7pCoi_Eigc4M/edit)

Once everything is checked off, you can go ahead and merge the PR. Don't forget to also delete the branch.

### CODE REVIEW

**Comment by ffinfo** [comment]
Forgot to mention that this also fixes a bug inside ThreadSafeProcessController.scala. Here the outputstreams are not closed. For small pipelines this is not a problem but was trying this on 60k jobs on a cluster with a limit of 65k file handles. Each job generate 2 streams so by pipeline did fail because I did reach the limit of 65k file handles ;).

File: public/gatk-queue/src/main/scala/org/broadinstitute/gatk/queue/QSettings.scala [hunk 1]
```diff
@@ -99,6 +100,13 @@ class QSettings {
   @Argument(fullName="log_directory", shortName="logDir", doc="Directory to write log files into.", required=false)
   var logDirectory: File = _
 
+  @Argument(fullName = "quality_of_serice", shortName = "qos", doc="Quality of Serice for Slurm", required=false)
```
Comment (vdauwera) [inline public/gatk-queue/src/main/scala/org/broadinstitute/gatk/queue/QSettings.scala]: double typo: "serice" should presumably be "service"?

File: public/gatk-queue/src/main/scala/org/broadinstitute/gatk/queue/QSettings.scala [hunk 2]
```diff
@@ -99,6 +100,13 @@ class QSettings {
...
+  @ClassType(classOf[String])
+  var qualityOfSerice: Option[String] = None
+
+  @Argument(fullName="waitBeforeJob", shortName="waitBeforeJob", doc="For Slurm the job submision waits number of seconds, incase of io delay on staorage. Default is 5 seconds", required=false)
```
Comment (vdauwera) [inline public/gatk-queue/src/main/scala/org/broadinstitute/gatk/queue/QSettings.scala]: clean up typos here too

**Review by vdauwera — COMMENTED** [review]
This seems useful for the community at large so in principle I think we'll be interested in this contribution, pending a full code review. However I'm not sure how quickly we can get someone to review this as we are short on resources and we have moved away from Queue for internal purposes. But I'll see what I can do. 

Note that a superficial level, I found a few typos that should be cleaned up.

File: public/gatk-queue/src/main/scala/org/broadinstitute/gatk/queue/QSettings.scala (hunk 1 above)
Comment (ffinfo) [inline public/gatk-queue/src/main/scala/org/broadinstitute/gatk/queue/QSettings.scala]: Fixed this

File: public/gatk-queue/src/main/scala/org/broadinstitute/gatk/queue/QSettings.scala (hunk 2 above)
Comment (ffinfo) [inline public/gatk-queue/src/main/scala/org/broadinstitute/gatk/queue/QSettings.scala]: Fixed this

**Comment by vdauwera** [comment]
Hi @ffinfo, sorry I never got back to you. We have been entirely focused on GATK4, and since we're no longer using Queue ourselves we're not going to be able to devote any resources to reviewing this. So I think the best thing to do is for you to continue development in your own fork independently of us.

**Pull request closed without merging**